### ISSUE TITLE
Does "heading in link" fail any WCAG success criterion?

### ISSUE THREAD

**Issue by radarxc** [w3c/wcag#2221]
Hi WCAG experts,

I have a question about heading nested in links(e.g., some html content like `<a href="path">...<h3>heading</h3></a>`).

So heading elements in links are commonly used on many websites, such as [Amazon search results](https://www.amazon.com/s?k=planter&crid=OM2PT3F3KCLW&sprefix=plan%2Caps%2C570&ref=nb_sb_noss_2) and [Bing search results](https://www.bing.com/search?q=cars&form=QBLH&sp=-1&pq=car&sc=5-3&qs=n&sk=&cvid=D746E9D592F34A99B0268E9BCA0658A7). But ~~after iOS 14~~ on iOS 14.x, VoiceOver ignores heading elements in link when users use heading navigation. Meanwhile, other screen readers(even MacOS VoiceOver) still read headings in links.

So is there any WCAG success criterion against headings in links? If there is some WCAG success criterions against this, is there any failure case for this?

Appreciate any suggestions in advance!

Update:
[patrickhlauke](https://github.com/patrickhlauke) tested iOS 15.3 VoiceOver([radarxc](https://github.com/radarxc) also tested and get the same results)
- Amazon: heading navigation works; linear navigation "omits to mention the headings.".
- Reduced test case: heading navigation and linear navigation work as expected.

**Comment by JAWS-test** [comment]
> Does "heading in link" fail any WCAG success criterion?

I would say not really. Because it's clearly a VoiceOver bug. According to the HTML specification, heading is allowed in links. If many screen readers would make the mistake, one could only argue with WCAG 5.2.4 (https://www.w3.org/TR/WCAG21/#cc4) that heading in links are not "accessibility supported". 
Regardless of whether WCAG allows or prohibits something, you should always try to make an app or page accessible, i.e. if VoiceOver has a bug, you should push to fix it or you should write your code so that the bug cannot occur (e.g. link in heading instead of heading in link).

**Comment by radarxc** [comment]
We already have workaround for this issue. I raised this issue here because I really would like to know if "headings in links" violate any rules.

Meanwhile, even it looks like VoiceOver behavior issue. But since iOS 14 released in 2020, many users reported this issue, and it still ignores headings in links. So I also want to know if we have any chance to make developers/companies aware of this issue, in case the ignoring on iOS is intended or need long time to fix. For this purpose, I think WCAG 5.2.4 maybe a good fit?

**Comment by patrickhlauke** [comment]
Apple often (always?) takes a "we know best" approach when choosing what to expose/support/which heuristics to apply. I remember how for the longest time they simply ignored/didn't do anything with `<fieldset>`/`<legend>` in terms of announcing that relationship in any way to their users. Pinning this on 5.2.4 (in light of the fact that on iOS there's no alternatives, so that effectively there is no other user agent/AT combo available to its users) would mean that audits would constantly have to flag things where VO has bugs or weird decisions on their part...which would not be the idea behind it.

Note also that VO does recognise these headings as heading - if you change the rotor to navigate by headings, you can jump between them as you should. It's only while navigating sequentially through the page that VO chooses not to mention the fact these are headings when it's already announcing them as links. So it's really more of a quirk/decision by VO than something that's "not accessibility supported".

**Comment by JAWS-test** [comment]
> An audit needs to hang a failure on an SC (and 5.2.4 could be part of that), but an audit can't just only point to 5.2.4.

If headings are not marked or not recognized as headings, this is a violation of SC 1.3.1

**Comment by patrickhlauke** [comment]
> If headings are not marked or not recognized as headings

but they are, and they are (if you navigate in VO by heading, it shows that it clearly knows those are headings inside the links). VO just chooses not to announce them as you navigate sequentially through the page.

**Comment by alastc** [comment]
Does it help if the markup is correctly nested, with the link inside the heading? 

I don't have a test page to hand, but VO can be particular about inline elements.

**Comment by patrickhlauke** [comment]
> Does it help if the markup is correctly nested, with the link inside the heading?

html5 allows links around block level elements

**Comment by patrickhlauke** [comment]
and to be clear, following some of the reasoning above, there's a risk of essentially using 5.2.4 to say "if authors code things in a perfectly valid way, but VO chooses not to announce it in one particular mode, then the authors fail" ... making SR accessibility de-factor "whatever VO chooses to do"

**Comment by radarxc** [comment]
> 

Yes, that a great point.

So since 5.2.4 does not fit for this case. Do we have any channel to let developers/companies(or even Apple?) to notice this known issue?

For now, 70% mobile screen reader users use VoiceOver. And this issue is a really impact for them.

**Comment by awkawk** [comment]
> to say "if authors code things in a perfectly valid way, but VO chooses not to announce it in one particular mode, then the authors fail" ... making SR accessibility de-facto "whatever VO chooses to do"

I'll start by saying that I think that this sounds like it is accessibility supported even in VO. AT don't always support features in the ways that developers/etc think that they should, but that doesn't mean that they aren't supported.

We dealt with this with ARIA in the days when ARIA support was emerging. Could you rely on aria to convey the role for a custom control in a way that met 4.1.2? Not always, and if accessibility support was nonexistent in the user agents that you based your conformance claim on, then you couldn't make a conformance claim that you met 4.1.2 with your custom tree element unless you had a conforming alternate version.

We also dealt with this with Flash and Silverlight - there were people that really wanted to say that as long as the correct information was exposed to MSAA that browsers and assistive technologies could use that the technique was conforming. But support was only on Windows with IE or FF, and the AT support was inconsistent, so end-users would definitely be affected if support wasn't in place.

If VO or NVDA or JAWS or a browser has an issue that interferes with meeting an SC, that issue can impact the conformance claim, but I don't think this rises to that level.

**Comment by radarxc** [comment]
> as said, when I tried on the Amazon search page, when switching rotor to heading navigation, i can quite happily jump between headings that are wrapped inside links, and VO does announce them as headings in that case. it's only when navigating sequentially through the page that VO omits to mention the headings.
>
> I just tried now with a reduced test case, just with
>
> ```<a href="#"><h1>Test</h1></a>```
> and here latest Safari/VO quite happily announces this as "Test, heading level 1, link"

It seems also work on my iPhone(I upgraded it yesterday)...Let me do more test and modify the original information to avoid misleading.

**Comment by fstrr** [comment]
This issue looks like it can be closed. If it needs to be re-opened, please start a new discussion.
